activation/activation.go

```diff
@@ -456,114 +395,118 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
 			return fmt.Errorf("failed to calculate activeset: %v", err)
 		}
 	}
-	if activeSetSize == 0 && !targetEpoch.IsGenesis() {
+	if activeSetSize == 0 && !(pubEpoch + 1).IsGenesis() {
 		return fmt.Errorf("empty active set size found! epochId: %v, len(view): %d, view: %v",
 			pubEpoch, len(view), view)
 	}
 
 	var commitment *types.PostProof
-	if b.prevATX == nil {
+	if b.challenge.PrevATXId == *types.EmptyAtxId {
 		commitment = b.commitment
 	}
 
-	atx := types.NewActivationTxWithChallenge(*b.challenge, b.getCoinbaseAccount(), activeSetSize, view, b.nipst, commitment)
+	atx := types.NewActivationTxWithChallenge(*b.challenge, b.getCoinbaseAccount(), activeSetSize, view, nipst, commitment)
 
-	b.log.With().Info("active ids seen for epoch", log.Uint64("pos_atx_epoch", uint64(posEpoch)),
+	b.log.With().Info("active ids seen for epoch", log.Uint64("atx_pub_epoch", uint64(pubEpoch)),
 		log.Uint32("view_cnt", activeSetSize))
 
-	signedAtx, err := b.SignAtx(atx)
-	if err != nil {
-		return err
-	}
-
-	buf, err := types.InterfaceToBytes(signedAtx)
-	if err != nil {
+	atxReceived := b.db.AwaitAtx(atx.Id())
+	if err := b.signAndBroadcast(atx); err != nil {
 		return err
 	}
-	b.prevATX = atx.ActivationTxHeader
-
-	// cleanup state
-	b.nipst = nil
-	b.challenge = nil
-	b.posLayerID = 0
 
-	err = b.net.Broadcast(AtxProtocol, buf)
-	if err != nil {
-		return err
-	}
+	b.log.Event().Info("atx published!",
+		log.AtxId(atx.ShortString()),
+		log.String("prev_atx_id", atx.PrevATXId.ShortString()),
+		log.String("post_atx_id", atx.PositioningAtx.ShortString()),
+		log.LayerId(uint64(atx.PubLayerIdx)),
+		log.EpochId(uint64(atx.PubLayerIdx.GetEpoch(b.layersPerEpoch))),
+		log.Uint32("active_set", atx.ActiveSetSize),
+		log.String("miner", b.nodeId.Key[:5]),
+		log.Int("view", len(atx.View)),
+	)
 
-	err = b.discardChallenge()
-	if err != nil {
-		log.Error("cannot discard Nipst challenge %s", err)
+	select {
+	case <-atxReceived:
+		b.log.Info("atx received in db")
+		events.Publish(events.AtxCreated{Created: true, Id: atx.ShortString(), Layer: uint64(b.currentEpoch())})
+	case <-b.layerClock.AwaitLayer((atx.TargetEpoch(b.layersPerEpoch) + 1).FirstLayer(b.layersPerEpoch)):
+		select {
+		case <-b.syncer.Await(): // ensure we've seen all blocks before concluding that the ATX was lost
+		case <-b.stop:
+			return &StopRequestedError{}
+		}
+		select {
+		case <-atxReceived:
```

Comment:
this channel should be unsubscribed from if calling stop since no one will read the data sent on it after stop, is there a unit test for these edge cases?

Reply:
This is not a problem because nobody ever sends data on the channel, it's only ever closed, which is non-blocking.

However, to avoid a leak (although it's tiny), I've implemented unsubscribing.